api/apps/api/src/modules/clone/import/domain/import/import.ts

```diff
@@ -0,0 +1,165 @@
+export type CompletePieceErrors =
+  | typeof componentNotFound
+  | typeof componentAlreadyCompleted;
+
+export class Import extends AggregateRoot {
+  private constructor(
+    private readonly id: string,
+    private readonly resourceId: string,
+    private readonly resourceKind: ResourceKind,
+    private readonly archiveLocation: ArchiveLocation,
+    private readonly pieces: ImportComponent[],
+  ) {
+    super();
+  }
+
+  static from(snapshot: ImportSnapshot): Import {
+    return new Import(
+      snapshot.id,
+      snapshot.resourceId,
+      snapshot.resourceKind,
+      snapshot.archiveLocation,
+      snapshot.importPieces.sort(this.orderSorter),
+    );
+  }
+
+  static new(data: Omit<ImportSnapshot, 'id'>): Import {
+    const id = v4();
+    const instance = new Import(
+      id,
+      data.resourceId,
+      data.resourceKind,
+      data.archiveLocation,
+      data.importPieces.sort(this.orderSorter),
+    );
+    instance.apply(new ImportRequested(id, data.resourceId, data.resourceKind));
+    instance.requestFirstBatch();
+    return instance;
+  }
+
+  completePiece(
+    piece: ImportComponentSnapshot &
+      Required<Pick<ImportComponentSnapshot, 'uri'>>,
+  ): Either<CompletePieceErrors, CompletePieceSuccess> {
+    const pieceToComplete = this.pieces.find(
+      (pc) => pc.id.value === piece.id.value,
+    );
+    if (!pieceToComplete) {
+      return left(componentNotFound);
+    }
+
+    if (pieceToComplete.uri) {
+      return left(componentAlreadyCompleted);
+    }
+
+    pieceToComplete.uri = piece.uri;
+
+    this.apply(
+      new PieceImported(
+        pieceToComplete.id,
+        pieceToComplete.piece,
+        pieceToComplete.resourceId,
+        pieceToComplete.uri,
+      ),
+    );
+
+    const nextBatch = this.getNextComponents(pieceToComplete);
+
+    if (nextBatch.finished) {
+      this.apply(new AllPiecesImported());
+    } else {
+      for (const nextPiece of nextBatch.components) {
+        this.apply(
+          new PieceImportRequested(
+            pieceToComplete.id,
+            pieceToComplete.piece,
+            pieceToComplete.resourceId,
+          ),
+        );
+      }
+    }
+
+    return right(true);
+  }
+
+  toSnapshot(): ImportSnapshot {
+    return {
+      id: this.id,
+      resourceId: this.resourceId,
+      resourceKind: this.resourceKind,
+      importPieces: this.pieces,
+      archiveLocation: this.archiveLocation,
+    };
+  }
+
+  private requestFirstBatch() {
+    if (this.pieces.length === 0) {
+      this.apply(new AllPiecesImported());
+    } else {
+      // we kindly assume that first batch is not done already
+      for (const component of this.pieces.filter(
+        (pc) => pc.order === this.pieces[0].order,
+      )) {
+        this.apply(
+          new PieceImportRequested(
+            component.id,
+            component.piece,
+            component.resourceId,
+          ),
+        );
+      }
+    }
+  }
+
+  private getNextComponents(
```

Comment:
Maybe this could be simplified. Something like:
```typescript
  private getNextComponents(
    lastPiece: ImportComponent,
  ): { components: ImportComponent[]; finished: boolean } {
    const elements = this.pieces.filter((pc) => pc.order === lastPiece.order);

    const somePieceDoesNotHaveUri = elements.some((pc) => !Boolean(pc.uri));

    if (somePieceDoesNotHaveUri)
      return {
        components: [],
        finished: false,
      };

    const location = this.pieces.indexOf(lastPiece);
    const nextComponentLocation = location + 1;
    const nextComponent = this.pieces[nextComponentLocation];

    return {
      components: this.pieces.filter((pc) => pc.order === nextComponent?.order),
      finished: !nextComponent,
    };
  }
```
Haven't properly checked this, but something like this with a guard clause seems clearer and should work.

Reply:
Thanks!

```
const elements = this.pieces.filter((pc) => pc.order === lastPiece.order);
    const hasIncompletePiecesForBatch = elements.some((pc) => !Boolean(pc.uri));

    if (hasIncompletePiecesForBatch) {
      return {
        components: [],
        finished: false,
      };
    }

    const location = this.pieces.indexOf(lastPiece);
    const nextComponentLocation = location + 1;
    const nextComponent = this.pieces[nextComponentLocation];

    if (nextComponent) {
      return {
        components: this.pieces.filter(
          (pc) => pc.order === nextComponent.order,
        ),
        finished: false,
      };
    }

    return {
      components: [],
      finished: true,
    };
  }
```

Ended with the above - tbh I feel that it is bit easier to understand the last piece with `if` clause instead shortening the code - but indeed the early return on not completed batch is good catch!

Reply:
updated in https://github.com/Vizzuality/marxan-cloud/pull/670/commits/5340b5e8fc1aadad9566620d8abd0adc67687f7f

Reply:
Nice! 💪🏽